### ISSUE TITLE
Enable LTCG for MSVC 14.0

### ISSUE THREAD

**Issue by raymond-w-ko** [vim/vim#496]
The version checks to map Visual Studio 2015 to MSVC 14.0 was added, but the checks for MSVC 14.0 were never done to determine that it was safe to enable link to code generation. Adding more a version check to the following conditionals fixes this and makes the binary generated by Visual Studio 2015 potentially faster.

``` diff
diff --git a/src/Make_mvc.mak b/src/Make_mvc.mak
index acc2ee5..dac756a 100644
--- a/src/Make_mvc.mak
+++ b/src/Make_mvc.mak
@@ -512,7 +512,7 @@ OPTFLAG = /O2
 OPTFLAG = /Ox
 !endif

-!if ("$(MSVCVER)" == "8.0") || ("$(MSVCVER)" == "9.0") || ("$(MSVCVER)" == "10.0") || ("$(MSVCVER)" == "11.0") || ("$(MSVCVER)" == "12.0")
+!if ("$(MSVCVER)" == "8.0") || ("$(MSVCVER)" == "9.0") || ("$(MSVCVER)" == "10.0") || ("$(MSVCVER)" == "11.0") || ("$(MSVCVER)" == "12.0") || ("$(MSVCVER)" == "14.0")
 # Use link time code generation if not worried about size
 !if "$(OPTIMIZE)" != "SPACE"
 OPTFLAG = $(OPTFLAG) /GL
@@ -1023,7 +1023,7 @@ LINKARGS2 = $(CON_LIB) $(GUI_LIB) $(NODEFAULTLIB) $(LIBC) $(OLE_LIB) user32.lib

 # Report link time code generation progress if used.
 !ifdef NODEBUG
-!if ("$(MSVCVER)" == "8.0") || ("$(MSVCVER)" == "9.0") || ("$(MSVCVER)" == "10.0") || ("$(MSVCVER)" == "11.0") || ("$(MSVCVER)" == "12.0")
+!if ("$(MSVCVER)" == "8.0") || ("$(MSVCVER)" == "9.0") || ("$(MSVCVER)" == "10.0") || ("$(MSVCVER)" == "11.0") || ("$(MSVCVER)" == "12.0") || ("$(MSVCVER)" == "14.0")
 !if "$(OPTIMIZE)" != "SPACE"
 LINKARGS1 = $(LINKARGS1) /LTCG:STATUS
 !endif
```

**Comment by k-takata** [comment]
If #224 is merged, this kind of patches would never be needed...

**Comment by brammool** [comment]
Raymond Ko wrote:

> The version checks to map Visual Studio 2015 to MSVC 14.0 was added,
> but the checks for MSVC 14.0 were never done to determine that it was
> safe to enable link to code generation. Adding more a version check to
> the following conditionals fixes this and makes the binary generated
> by Visual Studio 2015 potentially faster.

Can you try out the patch on issue 224?  It's a drastic change, I would
like to hear from a few people that it works properly.

## 

All true wisdom is found on T-shirts.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by raymond-w-ko** [comment]
I just tried it out with Visual Studio 2015 and everything works fine. /GL and /LTCG:STATUS were reported as part of the compiler and link flags in `:version` so the version detecting part is working correctly for me. I will be including this patch as part of my own branch.

I read through the patch and everything seems good. The only think I can think of is making the error message produced when version detection fails more helpful. Like changing:

```
+!message *** ERROR
+!message Cannot run Visual C to determine its version.
+!message You must have set up the compiler.
+!error Make aborted.
```

to something like:

```
+!message *** ERROR
+!message Cannot run Visual C to determine its version. Make sure cl.exe is in your PATH.
+!message This can usually be done by running the "vcvarsall.bat" file in VC directory where Visual Studio is installed.
+!error Make aborted.
```

**Comment by chrisbra** [comment]
Fixed with 7.4.960
